### ISSUE TITLE
[FSDP] Freezing some weights during training

### ISSUE THREAD

**Issue by QuentinDuval** [facebookresearch/fairscale#605]
## ❓ Freezing some weights during training

In the context of VISSL (SSL on Vision), we are using FSDP together with the pre-taining task SwAV.

SwAV pre-training requires us to maintain some weights frozen, while still flowing gradients through these weights: we train an encoder followed by a head, and some weights in the head (which we call the `prototypes`) need to be kept frozen for a few hundred of iterations at the beginning of the training to reach a good performance.

The way we do it is by setting `grad=None` on those weights for the first few hundred of iterations.
Is this something that is supported in FSDP?

### Motivation

In our work in integrating FSDP in VISSL, we have spent sometime making sure that the loss curves of FSDP and DDP match when training in FP32. In that context, we noticed that we can reach perfect loss curve overlap when we disable the `grad=None` on the `prototypes`. Here we have 4 runs, 2 with DDP, 2 with FSDP, all having the same curve:

<img width="739" alt="Screenshot 2021-04-09 at 16 23 14" src="https://user-images.githubusercontent.com/7412790/114731100-7486e580-9d0f-11eb-93a1-9fb9c3f50b1a.png">

But as soon as we enable this feature, the loss curves do not match anymore, the FSDP curve is more jittery and moves around the DDP curve:

<img width="727" alt="Screenshot 2021-04-09 at 16 23 53" src="https://user-images.githubusercontent.com/7412790/114731035-63d66f80-9d0f-11eb-8b31-27fed914385a.png">

We can reproduce this 100% of the time, so we suspect that setting `grad=None` is not interacting well with FSDP.

CC: @prigoyal @min-xu-ai @myleott

**Comment by min-xu-ai** [comment]
I don't see why it is not supported. Will add a unit test to ensure it is the same with ddp.

**Comment by prigoyal** [comment]
just wanted to chime in and mention: this issue is currently high-pri for us to figure out and we would appreciate any advice/help! :)

**Comment by min-xu-ai** [comment]
@QuentinDuval this issue is fixed with compute_dtype = fp32?

**Comment by prigoyal** [comment]
> @QuentinDuval this issue is fixed with compute_dtype = fp32?

yes, this is fixed by compute_dtype=fp32 :) I am running more tests and will comment later to close this issue.

**Comment by QuentinDuval** [comment]
After debugging, the issue was in VISSL, I get perfect convergence with the frozen prototypes now:

<img width="721" alt="Screenshot 2021-04-16 at 15 16 30" src="https://user-images.githubusercontent.com/7412790/115073773-c9209100-9ec6-11eb-9083-6f7e364b1b52.png">

Closing the issue.

**Comment by prigoyal** [comment]
Just connecting dots: this issue was due to the hardcoding of fsdp params names (as it doesn't return the original params names currently). Fixing this issue https://github.com/facebookresearch/fairscale/issues/605 would avoid such issues in future for VISSL or others.
